### PR TITLE
MID-2789: Allowlist for Annotation Passthrough

### PR DESCRIPTION
**Pull request by irishshagua**
* Add in annotation passthrough logic
  * Add testable annotations for playground
  * Format Code

Signed-off-by: bmooney <brian.mooney@zalando.ie>

### CODE REVIEW

File: src/test/scala/ie/zalando/fabric/gateway/service/IngressDerivationChainSpec.scala [hunk 1]
```diff
@@ -780,6 +802,18 @@ class IngressDerivationChainSpec extends FlatSpec with MockitoSugar with Matcher
...
 
+  "Annotation Handling" should "allow a subset of annotations through to the generated ingress" in {
+    val ingresses = testableRoutesDerivation(
+      sampleGateway,
+      GatewayMeta(DnsString.fromString("test-gateway").get, "default", None, Map("a" -> "1", "b" -> "2", "c" -> "3")))
+
+    ingresses.foreach { ingress =>
+      ingress.metadata.routeDefinition.additionalAnnotations.keySet should contain("a")
+      ingress.metadata.routeDefinition.additionalAnnotations.keySet should not contain("b")
+      ingress.metadata.routeDefinition.additionalAnnotations.keySet should not contain("c")
```
Comment (cjbooms) [inline src/test/scala/ie/zalando/fabric/gateway/service/IngressDerivationChainSpec.scala:813]: It's not immediately clear why a is allowed and b and c are not. I had to go digging through config to see the hard-coded test config. Is it possible to pass the allowlist into `testableRoutesDerivation` instead?
If not, change the test name to indicate that a is allowed:
``` 
"Annotation Handling" should "only allow the globally configured annotation with key `a`"
```
or rename the key from `a` to `ALLOWED`

**Comment by cjbooms** [comment]
It's hard to see the wood from the trees. Any chance you could pull out a separate pure reformatting/refactoring PR with updated logger config etc?

File: src/test/scala/ie/zalando/fabric/gateway/service/IngressDerivationChainSpec.scala (hunk 1 above)
Comment (irishshagua) [reply src/test/scala/ie/zalando/fabric/gateway/service/IngressDerivationChainSpec.scala]: Unfortunately the config is statically initialized on startup, so it's difficult to change in tests outside of the app properties file. I've updated the test name and variable names.

**Comment by irishshagua** [comment]
> It's hard to see the wood from the trees. Any chance you could pull out a separate pure reformatting/refactoring PR with updated logger config etc?

Ok formatter went to town on IngressDerivationChain. Will revert that. Outside of that the only other change was moving the tracing case class inside the AppConfig object which is pretty small.

**Comment by irishshagua** [comment]
👎

**Comment by irishshagua** [comment]
Added veto to ensure master branch is kept clean for BF

File: src/main/scala/ie/zalando/fabric/gateway/service/IngressDerivationChain.scala [hunk 2]
```diff
@@ -568,6 +575,13 @@ class IngressDerivationChain(stackSetOperations: StackSetOperations, versionedHo
           .mkString("{", ",", "}"))
   }
 
+  def filterAllowedAnnotations(meta: GatewayMeta, allowedAnnotationKeys: Set[String]): Map[String, String] = {
+    val allowedAnnotations = meta.annotations.filterKeys(allowedAnnotationKeys.contains)
+    log.debug(s"Annotations filter from ${meta.annotations} and passed through as $allowedAnnotations")
```
Comment (cjbooms) [inline src/main/scala/ie/zalando/fabric/gateway/service/IngressDerivationChain.scala:580]: Maybe we should change this to log a warn if annotations have been filtered out

Comment (irishshagua) [reply src/main/scala/ie/zalando/fabric/gateway/service/IngressDerivationChain.scala:580]: Annotations will always be filtered out. The likes of deployment-time etc. There's extra annotations that we receive from what people may actually add to their resource

**Comment by cjbooms** [comment]
👍

**Pull request closed without merging**